New Module: archive

bendoh opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
archive

##### ANSIBLE VERSION
```
ansible 2.2.0
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
This module serves two purposes:

* Allow management of compressed archives or single files using standard compression formats
* Act as an analog to the 'unarchive' module which does the opposite.

My particular use case was that I wanted to ensure a large SQL file stayed in a compressed state, and would update when the non-compressed version was restored. I initially opened a PR https://github.com/ansible/ansible-modules-core/pull/3735 against the file module to achieve this behavior, but it was clear that was not the right place for it.

This module offers the same functionality and more, by allowing multiple paths, trees, or patterns to be specified for archival. The common root for all of the files is computed and removed from the archived path names.

The simple use case is to compress a single file:

```yaml
- name: Compress this file
  archive: path=/my/file
```

When `/my/file` exists, `/my/file.gz` is created with the gzip-compressed contents of `/my/file`. If the `/my/file.gz` already exists, it is over-written and the difference in the compressed file size is used to determine whether or not the state has changed. If `/my/file.gz` exists and the source file `/my/file` does not, the state is considered `compress` and unchanged. `/my/file` is removed once compressed, *unless* `remove=False` is specified.

Another use:

```yaml
- name: Compress this tree
  archive: path=/my/tree creates=/my/archive.tgz
```

Creates a tarball at /my/archive.tgz with the contents of `/my/tree`. Since all of the files in the specified path share a common base path in `/my/tree`, the files in the produced archive are relative to `/my/tree`. The path `/my/tree` is removed unless `remove=False` is specified.

gregdek commented:
Thanks @bendoh for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

resmo commented on files/archive.py:
```diff
@@ -0,0 +1,360 @@
+  compression:
+    description:
+      - The type of compression to use. Can be 'gz', 'bz2', or 'zip'.
+    choices: [ 'gz', 'bz2', 'zip' ]
```
default is `gz` according arg spec.

bendoh replied:
Thanks! got it

abadger replied:
This conflates compression with archive format.  Probably do not want to do that.  Maybe:

```
choices: ['tar.gz', 'tar.bz2', 'zip']
```

Looking into this further, I also realized that this module attempts to guess whether the user wants to archive or compress.  We don't want to do that as the user may really require a single file tarball rather than a gz compressed file.  I'll discuss that in the issue rather than as a line comment.

Bikeshedding: the parameter name would be more accurately said to "format" or "archive_type" rather than compression.

abadger replied:
Also may want to support 'tar' (uncompressed tarfiles) here.

raben2 commented:
I was just testing the module. Seems good but i am from germany and if i run it on a directory in path
which contains vowel mutation Filenames it fails.
Maybe if you handle filenames as strings it could work.

bendoh commented:
Can you provide an example? Thanks for testing! Very helpful.

vazhnov commented:
I think «creates» must works like in module «[shell](http://docs.ansible.com/ansible/shell_module.html)».
So, please use another name for destination path, «dest» for example.

raben2 commented:
Sure no problem
my playbook

```
- hosts: localhost
  connection: local
  gather_facts: false
  become: false
  tasks:
    - archive:
        path: '/home/test/Downloads/*'
        creates: '/home/test/dl.tar.bz2'
        compression: bz2
      register: bzarchive
    - debug: msg="{{ bzarchive }}"
```

the debug output

```
Using /etc/ansible/ansible.cfg as config file
Loaded callback default of type stdout, v2.0

PLAYBOOK: test.yaml ************************************************************
1 plays in ./test.yaml

PLAY [localhost] ***************************************************************

TASK [archive] *****************************************************************
task path: /home/test/scripts/ansible-modules/test.yaml:6
Using module file /home/test/scripts/ansible-modules/ds/files/archive.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: test
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1464766245.33-145653721695337 `" && echo ansible-tmp-1464766245.33-145653721695337="` echo $HOME/.ansible/tmp/ansible-tmp-1464766245.33-145653721695337 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpco21zM TO /home/test/.ansible/tmp/ansible-tmp-1464766245.33-145653721695337/archive.py
<127.0.0.1> EXEC /bin/sh -c 'chmod -R u+x /home/test/.ansible/tmp/ansible-tmp-1464766245.33-145653721695337/ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'LANG=C LC_ALL=C LC_MESSAGES=C /usr/bin/python /home/test/.ansible/tmp/ansible-tmp-1464766245.33-145653721695337/archive.py; rm -rf "/home/test/.ansible/tmp/ansible-tmp-1464766245.33-145653721695337/" > /dev/null 2>&1 && sleep 0'
changed: [127.0.0.1] => {"archived": [], "arcroot": "/home/test/Downloads/", "changed": true, "creates": "/home/test/dl.tar.bz2", "expanded_paths": ["/home/test/Downloads/ArangoDB-2.8.9.tar.bz2", "/home/test/Downloads/content_pack.json", "/home/test/Downloads/virtualbox-5.0_5.0.20-106931~Ubuntu~wily_amd64.deb", "/home/test/Downloads/linux-headers-4.6.0-040600rc7-generic_4.6.0-040600rc7.201605081830_amd64.deb", "/home/test/Downloads/step1(1).png", "/home/test/Downloads/linux-headers-4.6.0-040600rc7_4.6.0-040600rc7.201605081830_all.deb", "/home/test/Downloads/a.png", "/home/test/Downloads/google-talkplugin_current_amd64.deb", "/home/test/Downloads/graylog-plugin-jira-master.zip", "/home/test/Downloads/graylog-plugin-jira-1.0.0.jar", "/home/test/Downloads/linux-image-4.6.0-040600rc7-generic_4.6.0-040600rc7.201605081830_amd64.deb", "/home/test/Downloads/über.JPG", "/home/test/Downloads/linux-firmware_1.157_all.deb", "/home/test/Downloads/step2.png", "/home/test/Downloads/3030068632725216.pdf", "/home/test/Downloads/Oracle_VM_VirtualBox_Extension_Pack-5.0.20-106931.vbox-extpack", "/home/test/Downloads/step1.png", "/home/test/Downloads/3030068638915006.pdf", "/home/test/Downloads/comonea-gunda-1.22.jar"], "invocation": {"module_args": {"backup": null, "compression": "bz2", "content": null, "creates": "/home/test/dl.tar.bz2", "delimiter": null, "directory_mode": null, "follow": false, "force": null, "group": null, "mode": null, "owner": null, "path": ["/home/test/Downloads/*"], "regexp": null, "remote_src": null, "remove": false, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null}, "module_name": "archive"}, "missing": [], "state": "archive"}

TASK [debug] *******************************************************************
task path: /home/test/scripts/ansible-modules/test.yaml:11
ok: [127.0.0.1] => {
    "msg": {
        "archived": [], 
        "arcroot": "/home/test/Downloads/", 
        "changed": true, 
        "creates": "/home/test/dl.tar.bz2", 
        "expanded_paths": [
            "/home/test/Downloads/ArangoDB-2.8.9.tar.bz2", 
            "/home/test/Downloads/content_pack.json", 
            "/home/test/Downloads/virtualbox-5.0_5.0.20-106931~Ubuntu~wily_amd64.deb", 
            "/home/test/Downloads/linux-headers-4.6.0-040600rc7-generic_4.6.0-040600rc7.201605081830_amd64.deb", 
            "/home/test/Downloads/step1(1).png", 
            "/home/test/Downloads/linux-headers-4.6.0-040600rc7_4.6.0-040600rc7.201605081830_all.deb", 
            "/home/test/Downloads/a.png", 
            "/home/test/Downloads/google-talkplugin_current_amd64.deb", 
            "/home/test/Downloads/graylog-plugin-jira-master.zip", 
            "/home/test/Downloads/graylog-plugin-jira-1.0.0.jar", 
            "/home/test/Downloads/linux-image-4.6.0-040600rc7-generic_4.6.0-040600rc7.201605081830_amd64.deb", 
            "/home/test/Downloads/über.JPG", 
            "/home/test/Downloads/linux-firmware_1.157_all.deb", 
            "/home/test/Downloads/step2.png", 
            "/home/test/Downloads/3030068632725216.pdf", 
            "/home/test/Downloads/Oracle_VM_VirtualBox_Extension_Pack-5.0.20-106931.vbox-extpack", 
            "/home/test/Downloads/step1.png", 
            "/home/test/Downloads/3030068638915006.pdf", 
            "/home/test/Downloads/comonea-gunda-1.22.jar"
        ], 
        "missing": [], 
        "state": "archive"
    }
}
```

the resulting file contains only 64 kb

raben2 commented:
I also would prefer to using dest instead of creates

bendoh commented:
Thanks for your review and testing! I've incorporated your feedback and feel pretty good about where it's at now. I've made the following changes
- creates -> dest input parameter
- ensure mutated vowel filenames work (though I don't think that particular bit was the bug, it was just failing to recurse properly)
- expand user home in dest

Here's my test tree:

```
 % ls -lR ~/vc/ansible-tests
/Users/bendoh/vc/ansible-tests:
total 1080
-rw-r--r-- 1 bendoh staff   6430 May 26 22:01 README.CR
-rw-r--r-- 1 bendoh staff   3155 May 26 22:01 USexport.msg
drwxr-xr-x 6 bendoh staff    204 May 31 23:30 ann
-rw-r--r-- 1 bendoh staff  20716 May 26 22:01 api.c
-rw-r--r-- 1 bendoh staff   7073 May 26 22:01 api.h
-rw-r--r-- 1 bendoh staff  38637 May 26 22:01 crc32.c
-rw-r--r-- 1 bendoh staff   1777 May 26 22:01 crc32.h
-rw-r--r-- 1 bendoh staff  12154 May 26 22:01 crc_i386.S
-rw-r--r-- 1 bendoh staff  22645 May 26 22:01 crypt.c
-rw-r--r-- 1 bendoh staff   4731 May 26 22:01 crypt.h
-rw-r--r-- 1 bendoh staff  34877 May 26 22:01 deflate.c
-rw-r--r-- 1 bendoh staff  17161 May 26 22:01 ebcdic.h
-rw-r--r-- 1 bendoh staff    557 May 26 22:01 file_id.diz
-rw-r--r-- 1 bendoh staff 141094 May 26 22:01 fileio.c
-rw-r--r-- 1 bendoh staff  11903 May 26 22:01 globals.c
-rw-r--r-- 1 bendoh staff  14169 May 26 22:01 match.S
drwxr-xr-x 4 bendoh staff    136 May 31 23:42 mütated
-rw-r--r-- 1 bendoh staff     10 Jun  1 08:57 playbook.retry
-rw-r--r-- 1 bendoh staff   1060 Jun  1 09:04 playbook.yml
-rw-r--r-- 1 bendoh staff   5992 May 26 22:01 revision.h
-rw-r--r-- 1 bendoh staff  20158 May 26 22:01 tailor.h
drwxr-xr-x 7 bendoh staff    238 May 31 23:30 text
-rw-r--r-- 1 bendoh staff  25824 May 26 22:01 timezone.c
-rw-r--r-- 1 bendoh staff   2831 May 26 22:01 timezone.h
-rw-r--r-- 1 bendoh staff  54437 May 26 22:01 trees.c
-rw-r--r-- 1 bendoh staff  20804 May 26 22:01 ttyio.c
-rw-r--r-- 1 bendoh staff   5367 May 26 22:01 ttyio.h
-rw-r--r-- 1 bendoh staff  41734 May 26 22:01 util.c
-rw-r--r-- 1 bendoh staff   2148 May 26 22:01 zbz2err.c
-rw-r--r-- 1 bendoh staff 193431 May 26 22:01 zip.c
-rw-r--r-- 1 bendoh staff  42139 May 26 22:01 zip.h
-rw-r--r-- 1 bendoh staff  23039 May 26 22:01 zipcloak.c
-rw-r--r-- 1 bendoh staff   4895 May 26 22:01 ziperr.h
-rw-r--r-- 1 bendoh staff 222430 May 26 22:01 zipfile.c
-rw-r--r-- 1 bendoh staff  27266 May 26 22:01 zipsplit.c

/Users/bendoh/vc/ansible-tests/ann:
total 20
-rw-r--r-- 1 bendoh staff 4372 May 26 22:01 zip30.ann
-rw-r--r-- 1 bendoh staff 3237 May 26 22:01 zip30f.ann
-rw-r--r-- 1 bendoh staff 1313 May 26 22:01 zip30g.ann
-rw-r--r-- 1 bendoh staff 1408 May 26 22:01 zip30h.ann

/Users/bendoh/vc/ansible-tests/mütated:
total 32
-rw-r--r-- 1 bendoh staff 22645 May 31 23:42 crypt.c
-rw-r--r-- 1 bendoh staff  4731 May 31 23:42 crypt.h

/Users/bendoh/vc/ansible-tests/text:
total 112
-rw-r--r-- 1 bendoh staff   642 May 26 22:01 Betas_Readme.txt
-rw-r--r-- 1 bendoh staff 97248 May 26 22:01 zip.txt
-rw-r--r-- 1 bendoh staff  2224 May 26 22:01 zipcloak.txt
-rw-r--r-- 1 bendoh staff  1909 May 26 22:01 zipnote.txt
```

And my test playbook

``` yaml
---
- hosts: all
  tasks:
    - name: Create archive.tgz of specified files
      archive:
        path: 
          - ~/vc/ansible-tests/api.*
          - ~/vc/ansible-tests/crypt.c
        dest: ~/vc/ansible-results/files.tgz

    - name: Create archive.tgz within an archive path
      archive:
        path: ~/vc/ansible-tests/
        dest: ~/vc/ansible-tests/archive.tgz

    - name: Create archive of mulitple directories
      archive:
        path:
          - ~/vc/ansible-tests/text
          - ~/vc/ansible-tests/ann
        dest: ~/vc/ansible-results/subdirs.tgz

    - name: Create archive of path with mutated vowels
      archive:
        path: ~/vc/ansible-tests/mütated
        dest: ~/vc/ansible-results/mutated.tgz

    - name: Create bz2 archive of path 
      archive:
        path: ~/vc/ansible-tests/ann
        dest: ~/vc/ansible-results/ann.tar.bz2

    - name: Create bz2 archive of path 
      archive:
        path: ~/vc/ansible-tests/text/*
        dest: ~/vc/ansible-results/text.tar.bz2
        compression: bz2
```

First run output:

```
>> ansible-playbook -i localhost, ~/vc/ansible-tests/playbook.yml -vv
No config file found; using defaults

PLAYBOOK: playbook.yml *********************************************************
1 plays in /Users/bendoh/vc/ansible-tests/playbook.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Create archive.tgz of specified files] ***********************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:4
changed: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/api.c", "/Users/bendoh/vc/ansible-tests/api.h", "/Users/bendoh/vc/ansible-tests/crypt.c"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": true, "dest": "/Users/bendoh/vc/ansible-results/files.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/api.c", "/Users/bendoh/vc/ansible-tests/api.h", "/Users/bendoh/vc/ansible-tests/crypt.c"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 13769, "state": "file", "uid": 501}

TASK [Create archive.tgz within an archive path] *******************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:11
changed: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/api.c", "/Users/bendoh/vc/ansible-tests/api.h", "/Users/bendoh/vc/ansible-tests/crc32.c", "/Users/bendoh/vc/ansible-tests/crc32.h", "/Users/bendoh/vc/ansible-tests/crc_i386.S", "/Users/bendoh/vc/ansible-tests/crypt.c", "/Users/bendoh/vc/ansible-tests/crypt.h", "/Users/bendoh/vc/ansible-tests/deflate.c", "/Users/bendoh/vc/ansible-tests/ebcdic.h", "/Users/bendoh/vc/ansible-tests/file_id.diz", "/Users/bendoh/vc/ansible-tests/fileio.c", "/Users/bendoh/vc/ansible-tests/globals.c", "/Users/bendoh/vc/ansible-tests/match.S", "/Users/bendoh/vc/ansible-tests/playbook.retry", "/Users/bendoh/vc/ansible-tests/playbook.yml", "/Users/bendoh/vc/ansible-tests/README.CR", "/Users/bendoh/vc/ansible-tests/revision.h", "/Users/bendoh/vc/ansible-tests/tailor.h", "/Users/bendoh/vc/ansible-tests/timezone.c", "/Users/bendoh/vc/ansible-tests/timezone.h", "/Users/bendoh/vc/ansible-tests/trees.c", "/Users/bendoh/vc/ansible-tests/ttyio.c", "/Users/bendoh/vc/ansible-tests/ttyio.h", "/Users/bendoh/vc/ansible-tests/USexport.msg", "/Users/bendoh/vc/ansible-tests/util.c", "/Users/bendoh/vc/ansible-tests/zbz2err.c", "/Users/bendoh/vc/ansible-tests/zip.c", "/Users/bendoh/vc/ansible-tests/zip.h", "/Users/bendoh/vc/ansible-tests/zipcloak.c", "/Users/bendoh/vc/ansible-tests/ziperr.h", "/Users/bendoh/vc/ansible-tests/zipfile.c", "/Users/bendoh/vc/ansible-tests/zipsplit.c", "/Users/bendoh/vc/ansible-tests/ann/zip30.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30f.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30g.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30h.ann", "/Users/bendoh/vc/ansible-tests/mütated/crypt.c", "/Users/bendoh/vc/ansible-tests/mütated/crypt.h", "/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": true, "dest": "/Users/bendoh/vc/ansible-tests/archive.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 313435, "state": "file", "uid": 501}

TASK [Create archive of mulitple directories] **********************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:16
changed: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt", "/Users/bendoh/vc/ansible-tests/ann/zip30.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30f.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30g.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30h.ann"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": true, "dest": "/Users/bendoh/vc/ansible-results/subdirs.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/text", "/Users/bendoh/vc/ansible-tests/ann"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 35617, "state": "file", "uid": 501}

TASK [Create archive of path with mutated vowels] ******************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:23
changed: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/mütated/crypt.c", "/Users/bendoh/vc/ansible-tests/mütated/crypt.h"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": true, "dest": "/Users/bendoh/vc/ansible-results/mutated.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/mütated"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 8416, "state": "file", "uid": 501}

TASK [Create bz2 archive of path] **********************************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:28
changed: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/ann/zip30.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30f.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30g.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30h.ann"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": true, "dest": "/Users/bendoh/vc/ansible-results/ann.tar.bz2", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/ann"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 4053, "state": "file", "uid": 501}

TASK [Create bz2 archive of path] **********************************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:33
changed: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt"], "arcroot": "/Users/bendoh/vc/ansible-tests/text/", "changed": true, "dest": "/Users/bendoh/vc/ansible-results/text.tar.bz2", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 26862, "state": "file", "uid": 501}

PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=6    unreachable=0    failed=0
```

Second run output:

```
 % ansible-playbook -i localhost, ~/vc/ansible-tests/playbook.yml -vv
No config file found; using defaults

PLAYBOOK: playbook.yml *********************************************************
1 plays in /Users/bendoh/vc/ansible-tests/playbook.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Create archive.tgz of specified files] ***********************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:4
ok: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/api.c", "/Users/bendoh/vc/ansible-tests/api.h", "/Users/bendoh/vc/ansible-tests/crypt.c"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": false, "dest": "/Users/bendoh/vc/ansible-results/files.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/api.c", "/Users/bendoh/vc/ansible-tests/api.h", "/Users/bendoh/vc/ansible-tests/crypt.c"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 13769, "state": "file", "uid": 501}

TASK [Create archive.tgz within an archive path] *******************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:11
ok: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/api.c", "/Users/bendoh/vc/ansible-tests/api.h", "/Users/bendoh/vc/ansible-tests/crc32.c", "/Users/bendoh/vc/ansible-tests/crc32.h", "/Users/bendoh/vc/ansible-tests/crc_i386.S", "/Users/bendoh/vc/ansible-tests/crypt.c", "/Users/bendoh/vc/ansible-tests/crypt.h", "/Users/bendoh/vc/ansible-tests/deflate.c", "/Users/bendoh/vc/ansible-tests/ebcdic.h", "/Users/bendoh/vc/ansible-tests/file_id.diz", "/Users/bendoh/vc/ansible-tests/fileio.c", "/Users/bendoh/vc/ansible-tests/globals.c", "/Users/bendoh/vc/ansible-tests/match.S", "/Users/bendoh/vc/ansible-tests/playbook.retry", "/Users/bendoh/vc/ansible-tests/playbook.yml", "/Users/bendoh/vc/ansible-tests/README.CR", "/Users/bendoh/vc/ansible-tests/revision.h", "/Users/bendoh/vc/ansible-tests/tailor.h", "/Users/bendoh/vc/ansible-tests/timezone.c", "/Users/bendoh/vc/ansible-tests/timezone.h", "/Users/bendoh/vc/ansible-tests/trees.c", "/Users/bendoh/vc/ansible-tests/ttyio.c", "/Users/bendoh/vc/ansible-tests/ttyio.h", "/Users/bendoh/vc/ansible-tests/USexport.msg", "/Users/bendoh/vc/ansible-tests/util.c", "/Users/bendoh/vc/ansible-tests/zbz2err.c", "/Users/bendoh/vc/ansible-tests/zip.c", "/Users/bendoh/vc/ansible-tests/zip.h", "/Users/bendoh/vc/ansible-tests/zipcloak.c", "/Users/bendoh/vc/ansible-tests/ziperr.h", "/Users/bendoh/vc/ansible-tests/zipfile.c", "/Users/bendoh/vc/ansible-tests/zipsplit.c", "/Users/bendoh/vc/ansible-tests/ann/zip30.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30f.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30g.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30h.ann", "/Users/bendoh/vc/ansible-tests/mütated/crypt.c", "/Users/bendoh/vc/ansible-tests/mütated/crypt.h", "/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": false, "dest": "/Users/bendoh/vc/ansible-tests/archive.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 313435, "state": "file", "uid": 501}

TASK [Create archive of mulitple directories] **********************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:16
ok: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt", "/Users/bendoh/vc/ansible-tests/ann/zip30.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30f.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30g.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30h.ann"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": false, "dest": "/Users/bendoh/vc/ansible-results/subdirs.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/text", "/Users/bendoh/vc/ansible-tests/ann"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 35617, "state": "file", "uid": 501}

TASK [Create archive of path with mutated vowels] ******************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:23
ok: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/mütated/crypt.c", "/Users/bendoh/vc/ansible-tests/mütated/crypt.h"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": false, "dest": "/Users/bendoh/vc/ansible-results/mutated.tgz", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/mütated"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 8416, "state": "file", "uid": 501}

TASK [Create bz2 archive of path] **********************************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:28
ok: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/ann/zip30.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30f.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30g.ann", "/Users/bendoh/vc/ansible-tests/ann/zip30h.ann"], "arcroot": "/Users/bendoh/vc/ansible-tests/", "changed": false, "dest": "/Users/bendoh/vc/ansible-results/ann.tar.bz2", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/ann"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 4053, "state": "file", "uid": 501}

TASK [Create bz2 archive of path] **********************************************
task path: /Users/bendoh/vc/ansible-tests/playbook.yml:33
ok: [localhost] => {"archived": ["/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt"], "arcroot": "/Users/bendoh/vc/ansible-tests/text/", "changed": false, "dest": "/Users/bendoh/vc/ansible-results/text.tar.bz2", "expanded_paths": ["/Users/bendoh/vc/ansible-tests/text/Betas_Readme.txt", "/Users/bendoh/vc/ansible-tests/text/zip.txt", "/Users/bendoh/vc/ansible-tests/text/zipcloak.txt", "/Users/bendoh/vc/ansible-tests/text/zipnote.txt", "/Users/bendoh/vc/ansible-tests/text/zipsplit.txt"], "gid": 20, "group": "staff", "missing": [], "mode": "0644", "owner": "bendoh", "size": 26862, "state": "file", "uid": 501}

PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0
```

Resulting archives:

```
62 Washbear ~/vc/ansible-results     ± 
 % for i in *; do echo "\n$i"; tar ztvf $i; done

ann.tar.bz2
-rw-r--r--  0 bendoh staff    4372 May 26 22:01 ann/zip30.ann
-rw-r--r--  0 bendoh staff    3237 May 26 22:01 ann/zip30f.ann
-rw-r--r--  0 bendoh staff    1313 May 26 22:01 ann/zip30g.ann
-rw-r--r--  0 bendoh staff    1408 May 26 22:01 ann/zip30h.ann

files.tgz
-rw-r--r--  0 bendoh staff   20716 May 26 22:01 api.c
-rw-r--r--  0 bendoh staff    7073 May 26 22:01 api.h
-rw-r--r--  0 bendoh staff   22645 May 26 22:01 crypt.c

mutated.tgz
-rw-r--r--  0 bendoh staff   22645 May 31 23:42 mütated/crypt.c
-rw-r--r--  0 bendoh staff    4731 May 31 23:42 mütated/crypt.h

subdirs.tgz
-rw-r--r--  0 bendoh staff     642 May 26 22:01 text/Betas_Readme.txt
-rw-r--r--  0 bendoh staff   97248 May 26 22:01 text/zip.txt
-rw-r--r--  0 bendoh staff    2224 May 26 22:01 text/zipcloak.txt
-rw-r--r--  0 bendoh staff    1909 May 26 22:01 text/zipnote.txt
-rw-r--r--  0 bendoh staff    1269 May 26 22:01 text/zipsplit.txt
-rw-r--r--  0 bendoh staff    4372 May 26 22:01 ann/zip30.ann
-rw-r--r--  0 bendoh staff    3237 May 26 22:01 ann/zip30f.ann
-rw-r--r--  0 bendoh staff    1313 May 26 22:01 ann/zip30g.ann
-rw-r--r--  0 bendoh staff    1408 May 26 22:01 ann/zip30h.ann

text.tar.bz2
-rw-r--r--  0 bendoh staff     642 May 26 22:01 Betas_Readme.txt
-rw-r--r--  0 bendoh staff   97248 May 26 22:01 zip.txt
-rw-r--r--  0 bendoh staff    2224 May 26 22:01 zipcloak.txt
-rw-r--r--  0 bendoh staff    1909 May 26 22:01 zipnote.txt
-rw-r--r--  0 bendoh staff    1269 May 26 22:01 zipsplit.txt

 % tar ztvf ~/vc/ansible-tests/archive.tgz
drwxr-xr-x  0 bendoh staff       0 May 31 23:30 ann/
drwxr-xr-x  0 bendoh staff       0 May 31 23:42 mütated/
drwxr-xr-x  0 bendoh staff       0 May 31 23:30 text/
-rw-r--r--  0 bendoh staff   20716 May 26 22:01 api.c
-rw-r--r--  0 bendoh staff    7073 May 26 22:01 api.h
-rw-r--r--  0 bendoh staff   38637 May 26 22:01 crc32.c
-rw-r--r--  0 bendoh staff    1777 May 26 22:01 crc32.h
-rw-r--r--  0 bendoh staff   12154 May 26 22:01 crc_i386.S
-rw-r--r--  0 bendoh staff   22645 May 26 22:01 crypt.c
-rw-r--r--  0 bendoh staff    4731 May 26 22:01 crypt.h
-rw-r--r--  0 bendoh staff   34877 May 26 22:01 deflate.c
-rw-r--r--  0 bendoh staff   17161 May 26 22:01 ebcdic.h
-rw-r--r--  0 bendoh staff     557 May 26 22:01 file_id.diz
-rw-r--r--  0 bendoh staff  141094 May 26 22:01 fileio.c
-rw-r--r--  0 bendoh staff   11903 May 26 22:01 globals.c
-rw-r--r--  0 bendoh staff   14169 May 26 22:01 match.S
-rw-r--r--  0 bendoh staff      10 Jun  1 08:57 playbook.retry
-rw-r--r--  0 bendoh staff    1060 Jun  1 09:04 playbook.yml
-rw-r--r--  0 bendoh staff    6430 May 26 22:01 README.CR
-rw-r--r--  0 bendoh staff    5992 May 26 22:01 revision.h
-rw-r--r--  0 bendoh staff   20158 May 26 22:01 tailor.h
-rw-r--r--  0 bendoh staff   25824 May 26 22:01 timezone.c
-rw-r--r--  0 bendoh staff    2831 May 26 22:01 timezone.h
-rw-r--r--  0 bendoh staff   54437 May 26 22:01 trees.c
-rw-r--r--  0 bendoh staff   20804 May 26 22:01 ttyio.c
-rw-r--r--  0 bendoh staff    5367 May 26 22:01 ttyio.h
-rw-r--r--  0 bendoh staff    3155 May 26 22:01 USexport.msg
-rw-r--r--  0 bendoh staff   41734 May 26 22:01 util.c
-rw-r--r--  0 bendoh staff    2148 May 26 22:01 zbz2err.c
-rw-r--r--  0 bendoh staff  193431 May 26 22:01 zip.c
-rw-r--r--  0 bendoh staff   42139 May 26 22:01 zip.h
-rw-r--r--  0 bendoh staff   23039 May 26 22:01 zipcloak.c
-rw-r--r--  0 bendoh staff    4895 May 26 22:01 ziperr.h
-rw-r--r--  0 bendoh staff  222430 May 26 22:01 zipfile.c
-rw-r--r--  0 bendoh staff   27266 May 26 22:01 zipsplit.c
-rw-r--r--  0 bendoh staff    4372 May 26 22:01 ann/zip30.ann
-rw-r--r--  0 bendoh staff    3237 May 26 22:01 ann/zip30f.ann
-rw-r--r--  0 bendoh staff    1313 May 26 22:01 ann/zip30g.ann
-rw-r--r--  0 bendoh staff    1408 May 26 22:01 ann/zip30h.ann
-rw-r--r--  0 bendoh staff   22645 May 31 23:42 mütated/crypt.c
-rw-r--r--  0 bendoh staff    4731 May 31 23:42 mütated/crypt.h
-rw-r--r--  0 bendoh staff     642 May 26 22:01 text/Betas_Readme.txt
-rw-r--r--  0 bendoh staff   97248 May 26 22:01 text/zip.txt
-rw-r--r--  0 bendoh staff    2224 May 26 22:01 text/zipcloak.txt
-rw-r--r--  0 bendoh staff    1909 May 26 22:01 text/zipnote.txt
-rw-r--r--  0 bendoh staff    1269 May 26 22:01 text/zipsplit.txt
```

raben2 commented:
works_for_me
shipit
thanks!

dagwieers commented:
I like the fact that this is a pure-python module.

Now FYI, unarchive is not a pure-python module and will never support check-mode properly (since we are using gtar). And the general idea is to reimplement unarchive as an idempotent pure-python module using a different name (i.e. uncompress) as we can never ensure a pure-python implementation is backward compatible (because of extra-opts and gtar/unzip specifics).

bendoh commented:
Should unarchive be re-implemented as 'uncompress', I suppose this module would then be renamed to 'compress' just to make the association clear. 

I don't think it would be too much of a stretch to make this thing work backwards as 'uncompress' in another PR.

dagwieers commented:
@bendoh Agreed, not a showstopper, just a heads up.

shipit

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+        argument_spec = dict(
+            path = dict(type='list', required=True),
+            compression = dict(choices=['gz', 'bz2', 'zip'], default='gz', required=False),
+            dest = dict(required=False),
```
This should probably be type='path'

alikins commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+        path = os.path.expanduser(path)
+
+        # Detect glob-like characters
+        if any((c in set('*?')) for c in path):
```
any() is a python 2.5+ method

modules target python2.4 unless a higher version is specified in the Requirements: field

abadger replied:
This is probably quicker written as:

``` python
if ('*' in path or '?' in path):
```

The code you have is looping over every character in the path, creating a set each time, and then doing the containment test.  Simply checking for containment twice is probably faster.

abadger replied:
Alternately, if speed isn't an issue here, you could probably just run glob.glob(path) on every path and not worry about prechecking for glob characters.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+    successes = []
+
+    for i, path in enumerate(paths):
+        path = os.path.expanduser(path)
```
Most of our paths we expand env vars as well so this should probably be:

``` python
path = os.path.expanduser(os.path.expandvars(path))
```

alikins commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+    # No source files were found but the named archive exists: are we 'compress' or 'archive' now?
+    if len(missing) == len(expanded_paths) and dest and os.path.exists(dest):
+        # Just check the filename to know if it's an archive or simple compressed file
+        if re.search(r'(\.tar\.gz|\.tgz|.tbz2|\.tar\.bz2|\.zip)$', os.path.basename(dest), re.IGNORECASE):
```
I'd add an 'import re'

The 're' module gets imported into this namespace via the module_utils.basic \* include, but would prefer the
explicit import.

alikins commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+
+    archive_paths = []
+    missing = []
+    exclude = []
```
exclude appears to never be used

alikins commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+
+import stat
+import os
+import errno
```
errno and stat don't appear to be used.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+            expanded_paths = expanded_paths + glob.glob(path)
+        else:
+            expanded_paths.append(path)
+
```
Note: There's a difference between these two code paths.  Simply appending the path does not check if the path exists.  glob.glob() will return an empty list if no files and directories matched the path.

bendoh replied:
Worth a comment!

alikins commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+                        elif compression == 'bz2':
+                            f_out = bz2.BZ2File(dest, 'wb')
+                        else:
+                            raise OSError("Invalid compression")
```
This could be detected earlier and avoid the slightly odd use of OSError here (presumably to trigger the except handler below?).

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+            expanded_paths.append(path)
+
+    if len(expanded_paths) == 0:
+        return module.fail_json(path=', '.join(paths), expanded_paths=', '.join(expanded_paths), msg='Error, no source paths were found')
```
The previous comment about expanded_paths.append() vs expanded_path + glob.glob() comes into play here.  If the user gave one or more paths that don't exist on the system and do not have glob characters then this failure won't be triggered.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+    # Default created file name (for single-file archives) to
+    # <file>.<compression>
+    if dest:
+        dest = os.path.expanduser(dest)
```
If we list dest as type=path in the srgument spec we don't have to expanduser() here.

gundalow commented:
@bendoh A few review comments came in after I merged this. Could you please raise a fresh PR to fix those issues?

Thanks for this new module. We like the pure Python implementation.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+
+    # If we actually matched multiple files or TRIED to, then
+    # treat this as a multi-file archive
+    archive = globby or os.path.isdir(expanded_paths[0]) or len(expanded_paths) > 1
```
globby is always False here (globby is set to an initial value at the top of the main() function but then never changed.  Perhaps it just needs to be removed altogether?)

bendoh replied:
It should get set to true when a glob character is detected; so even only one file matches the glob, it will generate an archive instead of a simple compressed file. I'll update that in my future PR...

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+module: archive
+version_added: 2.2
+short_description: Creates a compressed archive of one or more files or trees.
+extends_documentation_fragment: files
```
Actually... I don't see support for notable file module parameters like mode down below...  Documenting that this supports the file module arguments is overpromising what the module can do so we should not use this unless we also support mode, owner, group, and selinux settings. (These are easier than in the unarchive module as they'd be applied to the tar/zip file but they aren't implemented yet.)

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+            dest = dict(required=False),
+            remove = dict(required=False, default=False, type='bool'),
+        ),
+        add_file_common_args=True,
```
See my note about extends_documentation_fragment above.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+- archive: path=/path/to/foo dest=/path/to/foo.tgz
+
+# Compress regular file /path/to/foo into /path/to/foo.gz and remove it
+- archive: path=/path/to/foo remove=True
```
i'll make a toplevel comment about compression vs archiving.  Will need to change this if we move compression to a different module.

bendoh replied:
See my earlier comment on this PR, I don't feel that a distinction is truly warranted, as from the average user's perspective, they are the same thing. Definitely worth discussion, though.

bendoh commented:
Thanks for the feedback, I'll open a new PR with fixes soon.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+            if i < len(arcroot):
+                arcroot = os.path.dirname(arcroot[0:i+1])
+
+            arcroot += os.sep
```
xrange is not python3 compatible.  Not sure this does the right thing if, for example, you have `expanded_paths = ['/srv/one-two/', '/srv/one-three']`

I think it will end up with `arcroot = '/srv/one-t/'`.  Probably need to write some code based around recursively calling `os.path.split()` or (a little hackier but usually correct) `arcroot.split(os.path.sep)`

bendoh replied:
Except that `os.path.dirname` is used to determine the actual shared parent directory, it's not just extracting substrings from the path!

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+
+            state = 'compress'
+
+        if remove:
```
This isn't check_mode protected.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+                # Easier compression using tarfile module
+                elif compression == 'gz' or compression == 'bz2':
+                    arcfile = tarfile.open(dest, 'w|' + compression)
+
```
Creating of the above two files are not prevented in check_mode

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+            if len(errors) > 0:
+                module.fail_json(msg='Errors when writing archive at %s: %s' % (dest, '; '.join(errors)))
+
+        if state in ['archive', 'incomplete'] and remove:
```
Removal of files isn't prevented in check_mode.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+            remove = dict(required=False, default=False, type='bool'),
+        ),
+        add_file_common_args=True,
+        supports_check_mode=True,
```
Reviewing other code, I see that check_mode is _very_ badly broken right now... May want to quickly remove supports_check_mode until you can be sure that those are fixed.  We don't want people removing their files when they think that their remote systems won't be touched.

abadger commented on files/archive.py:
```diff
@@ -0,0 +1,385 @@
+    module.exit_json(archived=successes, dest=dest, changed=changed, state=state, arcroot=arcroot, missing=missing, expanded_paths=expanded_paths)
+
+# import module snippets
+from ansible.module_utils.basic import *
```
Style note: This import can go near the top with the other imports now.  It no longer needs to be a wildcard import -- you can import exactly what you need (for instance, from ansible.moudle_utils.basic import AnsibleModule)

abadger commented:
Couple toplevel things to note as well:
- style-wise, splitting the one big main() function into separate functions will make it easier to read.  Eventually we'll want to unittest modules and separating code which has side effects from code which doesn't will make that easier.  But this module has a lot of code that is side effect driven so we may need to leave that until later.
- This module mixes compression with archiving and attempts to guess whether the user meant one or the other.  This isn't a good idea as a ser may want to place a single file into a tar.gz.  Couple options:
  - We could make the choice explicit (for instance, if we changed compression to list both tar.gz and gz formats, the code could then recursively gzip all the files it found in the specified paths).
  - We could leave this as archiving only and push uncompressing to its own module.

Right now I favour the latter for a few reasons:
- Leaving compression out of this module will mirror what unarchive does.
- I don't see us adding raw uncompress support to the unarchive module, at least anytime soon. unarchive is pretty complex as it is and compression would be an added thing to worry about.
- There's a module review for uncompress ( https://github.com/ansible/ansible-modules-extras/pull/1301  and https://github.com/ansible/ansible/pull/13352 ) that just needs a few more wrinkles ironed out.  So a compress module will match with that.

So rather than spending too much time trying to figure out how best to change the UI here to handle both archives and simple compressors, we should probably create a second module that handles pure compression and take it out of this one.

dagwieers commented:
To be honest, I don't think making a distinction between **unarchive** and **uncompress** makes a lot of sense to our users. In most (if not all) cases, they want to get the content and not the archive out of a compressed file (if they already understand the difference between an archive and a compressed file, as zip does not make this distinction).

My plan was to deprecate **unarchive**, and replace it with a pure-python **uncompress** module since we cannot create a pure-python unarchive module as a drop-in replacement anyhow (extra_opts simply ceases to exist so it breaks stuff). 

To summarize, it means that this module would become **compress** instead, and I welcome this contribution !

bendoh commented:
@abadger Thank you for your feedback. I haven't done a lot of work with Python, so I followed the structure of other modules as examples until I felt a bit more comfortable. I'm not too familiar with Python2 vs 3 compatibility issues, either, so thanks for pointing them out. I will work on an update tonight and submit another PR.

I like the idea of having more explicit output format options, so we know exactly what kind of files we get out.  In terms of the idea you raise of recursively compressing files within a tree with gzip, I think it would add a bulky chunk of code, but it is definitely worth exploring.

@dagwieers I agree that many users wouldn't really make the distinction between a _compressed_ and an _archived_ file, and it seems like splitting hairs to me. This module attempts to smooth over the differences that do exist between those concepts in a predictable way.

For parity's sake, however, since you intend to deprecate _unarchive_ in favor of the pure-python _uncompress_, wouldn't it be logical to rename this module to _compress_? That way the _archive_/_unarchive_ commands would be logically grouped (as unavailable/deprecated) and the _compress_/_uncompress_ commands would be similarly grouped, as two pure-python modules. It would then become the practice to use that pair of verbs to describe the concept of "squishing file(s)", and be more coherent to users. The reason I chose _archive_ as the name in the first place was to be an obvious opposite of the existing _unarchive_ command.

dagwieers commented:
@bendoh That would indeed have my preference, but I don't think my plan was ever vetted by _the powers that be_ ;-) There is some backstory in ansible/ansible-modules-core#3307

BTW I think your **archive** (or rather **compress**) module ought to live in ansible-modules-core, next to **unarchive** (or preferably the new/future pure-python **uncompress**).

abadger commented:
I've added the archive+compress question to the core meeting agenda so we can get an overall plan sorted out before 2.2.

abadger commented:
@bendoh @dagwieers We talked about it at yesterdays meeting and it seems that we're okay with either single module or two modules.  However we're concerned about the parameters and how they will affect what the user can do.  We're not sure if we have time to review the parameters for 2.2 so we're going to debate what our options for 2.2 are at the meeting tomorrow.

dagwieers commented:
@abadger Thanks for the heads-up.

This is related to the specifics of archive/compress, but I would also be interested to know what the plan is for a pure-python uncompress module (and the deprecation of unarchive in favor of it). There are some restrictions if we go pure-python, but we would end up with a first-class idempotent and check-mode/diff-mode supporting module. (Not the current bug-ridden implementation that scrapes output :-))

abadger commented:
@dagwieers <nod>  I don't think it's on our radar to write such a thing ourselves but we'd certainly consider deprecating unarchive for such a module if it was contributed.

Regarding archive, at the meeting yesterday we came to the conclusion that no one on the core team has the time to go over the parameters before Feature Freeze.  Time may or may not free up after feature freeze and before the 2.2 release candidates start (with many new features come many new bugs :-(  We decided that since users rely on the parameters staying the same (or being replaced in a backwards compatible manner) we don't want to ship archive in 2.2 without getting those right.  So once stable-2.2 branches, we'll revert archive in that branch but leave it in devel for 2.3.  If someone is able to review the parameters and come up with a new design that handles the cornercases (single file archives and multi-file compression) and then we're able to merge code to implement that before the 2.2 rc's start then we can cherry-pick the changes back into 2.2.  If not, we'll make every effort to get that done for 2.3.

I'm very sorry that we don't have more time to spend on giving guidance about the parameters now.  Thanks for your patience.

bendoh commented:
I'm just excited to have been able to contribute! Thanks @abadger, look forward to getting this into 2.3.  There are a couple of things I need to clean up for old-python compatibility, in any case. And thank you @dagwieers for all your input.
